# UniVH

## Overview
RNA viruses, characterized by their exceptionally high mutation rates and rapid adaptability, pose significant challenges to public health. Despite the vast amounts of viral sequence data revealed through metagenomic sequencing, identifying viruses with high zoonotic potential remains a critical issue in pandemic preparedness. Existing host prediction models often suffer from limitations in accuracy and coverage, particularly when applied to emerging viruses and rare host species.

To overcome the scalability and precision limitations of current methods, we developed the UniVH framework. This unified framework simultaneously models associations between different viral families and host taxa, filling the gaps in both accuracy and coverage. By leveraging advanced genomic language models, UniVH embeds the complete gene repertoires of viruses and hosts into a unified representation space, efficiently compressing viral and host features based on protein functional classification to improve prediction accuracy.

UniVH was trained on a large-scale dataset, encompassing 90 viral families and 240 host families, enabling species-level host predictions for emerging and under-researched viruses.

---

## Requirements
- Python packages:
  - `torch==1.12.1.post201`
  - `pandas==2.2.3`
  - `numpy==2.2.4`
  - `matplotlib==3.10.1`
  - `sklearn==1.6.1`

---

## Data Preparation
1. **Protein Functional Annotation**  
   - For the viruses to be predicted, annotate their proteins using **HMM**.  
   - Simultaneously, generate a CSV file containing all possible virus-host combinations for these viruses.
   - To perform the annotation, run the following command:
     ```bash
     python 01_hmmscan.py
     ```

2. **Protein Embedding**  
   - Use the protein language model **[LucaOne](https://github.com/LucaOne/LucaOne)** to calculate embeddings for each protein of the viruses to be predicted.
   - To generate the protein embeddings, execute the following script:
     ```bash
     bash 01_lucaone.sh
     ```

3. **Dataset Access**  
   - The required data and pre-trained model files are available on **[Zenodo](https://doi.org/10.5281/zenodo.17505409)**.

---

## Usage
- During prediction, provide the nucleotide and amino acid sequences of the virus CDS (coding sequence).  
  - **Nucleotide sequences** are represented using embeddings generated by **LucaOne**.  
  - **Amino acid sequences** are functionally annotated via **HMM**.

- To run the prediction, execute the following command:
  ```bash
  python 02_predict_geoNA.py
  ```

- The `test` folder contains an example of coronavirus prediction to illustrate the workflow.

---

## Reference
If you use this dataset or framework, please cite the corresponding publication.







